### ISSUE TITLE
Issue with far pointers to struct with embedded strings

### ISSUE THREAD

**Issue by bartoldeman** [tkchia/gcc-ia16#2]
Hi,

I am playing a little bit with your far pointer support and the freedos kernel. It's getting there, great work, with a few changes but it's modulo bugs. I extracted a minimal example for one of the bugs (compile with -Os):

```
struct string_int {
  char str[1];
  int number;
};

int p_number_not_zero(struct string_int __far *p)
{
  return p->number != 0;
}
```

```
$ ia16-elf-gcc -Os -c tst.c
tst.c: In function ‘p_number_not_zero’:
tst.c:9:1: error: unrecognizable insn:
 }
 ^
(insn 10 9 11 2 (set (subreg:HI (reg:SI 29) 0)
        (plus:HI (plus:HI (plus:HI (reg:HI 26)
                    (unspec:HI [
                            (reg:HI 27)
                        ] 1))
                (const_int 2 [0x2]))
            (const_int 2 [0x2]))) tst.c:8 -1
     (nil))
tst.c:9:1: internal compiler error: in extract_insn, at recog.c:2287
0x852e58f _fatal_insn(char const*, rtx_def const*, char const*, int, char const*)
        ../../gcc-ia16/gcc/rtl-error.c:108
0x852e5cc _fatal_insn_not_found(rtx_def const*, char const*, int, char const*)
        ../../gcc-ia16/gcc/rtl-error.c:116
0x8506f6d extract_insn(rtx_insn*)
        ../../gcc-ia16/gcc/recog.c:2287
0x8363619 instantiate_virtual_regs_in_insn
        ../../gcc-ia16/gcc/function.c:1582
0x8363619 instantiate_virtual_regs
        ../../gcc-ia16/gcc/function.c:1950
0x8363619 execute
        ../../gcc-ia16/gcc/function.c:1999
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <http://gcc.gnu.org/bugs.html> for instructions.
```

It compiles ok without -Os and also if the str field is removed.

**Comment by tkchia** [comment]
Hello @bartoldeman,

Thanks!  I have reproduced the bug and will try to look into it (and hopefully learn yet more about GCC internals in the process).

One possible workaround in the meantime is to use far pointers to `volatile` things (e.g. `volatile struct string_int __far *`).  This should work at all optimization levels.

**Comment by tkchia** [comment]
Hello @bartoldeman,

This bug (and bug #5) should be fixed now in commit `0a29430437`.  Sorry about the previous failed commits --- I made some errors in my previous commit messages.

Thank you!

**Comment by bartoldeman** [comment]
Thanks for fixing. I'm getting a lot further now especially with the workaround in bug #3.
